### PR TITLE
Implemented some live map stats tracking.

### PR DESCRIPTION
**Pull request by Brollyy**
Implemented some live map stats tracking.

I've elected to drop CPU usage and Effective BPM, sincethey don't quite fit into UI on window mode and I'm not sure they'd deliver that much value over what's been done. Adding them later shouldn't be an issue, although CPU usage is not really a map statistic, so I don't think it should be in that section to begin with.

image
image

### CODE REVIEW

File: Classes/MapEditor/Stats/MapStats.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                } else {
+                    lineNoteCount++;
+                }
+            }
+            // last row to consider after all the notes were processed
+            if (lineNoteCount == rowCount) {
+                rowNoteCount++;
+            }
+
+            return rowNoteCount;
```
Comment (Nytilde) [inline Classes/MapEditor/Stats/MapStats.cs:77]: maybe just 

```csharp
return notes.GroupBy(note => note.beat)
    .Select(group => group.Count())
    .Count(count => count == rowCount);
```
should be sufficient

File: Classes/MapEditor/Stats/MapStats.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            var noteIndex = 0;
+
+            while (noteIt.MoveNext()) {
+                while (Helper.DoubleApproxGreater(noteIt.Current.beat - startIt.Current.beat, beats)) {
+                    startIt.MoveNext();
+                    startIndex++;
+                }
+                peakNPB = Math.Max(peakNPB, (noteIndex - startIndex + 1) / beats);
+                noteIndex++;
+            }
```
Comment (Nytilde) [inline Classes/MapEditor/Stats/MapStats.cs:110]: i would prefer to do this immutable:
```csharp
var peakNPB = notes
    .Select((note, noteIndex) => notes
        .TakeWhile((_, startIndex) => Helper.DoubleApproxGreater(note.beat - notes.ElementAt(startIndex).beat, beats))
        .Count())
    .Select((noteCount, noteIndex) => (noteIndex - noteCount + 1) / beats)
    .DefaultIfEmpty(0)
    .Max();
```

File: Classes/MapEditor/Stats/MapStats.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        public void RecalculateNotes(SortedSet<Note> notes, SortedSet<Note> selectedNotes) {
+            this.allNotes = notes.Count;
+            this.selectedNotes = selectedNotes.Count;
+            this.singleNotes = CalculateRowNotes(notes, 1);
+            this.doubleNotes = CalculateRowNotes(notes, 2);
+        }
+
+        private int CalculateRowNotes(SortedSet<Note> notes, int rowCount) {
```
Comment (Nytilde) [inline Classes/MapEditor/Stats/MapStats.cs:54]: static

File: Classes/MapEditor/Stats/MapStats.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            while (noteIt.MoveNext()) {
+                while (Helper.DoubleApproxGreater(noteIt.Current.beat - startIt.Current.beat, beats)) {
+                    startIt.MoveNext();
+                    startIndex++;
+                }
+                peakNPB = Math.Max(peakNPB, (noteIndex - startIndex + 1) / beats);
+                noteIndex++;
+            }
+
+            return double.Round(peakNPB * this.globalBPM / 60, 2);
```
Comment (Nytilde) [inline Classes/MapEditor/Stats/MapStats.cs:112]:
```csharp
return double.Round(peakNPB * globalBPM / 60, 2);
```
`this` is not required

**Review by Nytilde — CHANGES_REQUESTED** [review]
maybe a few small changes, but overall i think this is a great new feature :)

File: Classes/MapEditor/Stats/MapStats.cs (hunk 1 above)
Comment (Brollyy) [reply Classes/MapEditor/Stats/MapStats.cs:77]: Needed some changes to make sure floating-point rounding didn't throw off the count, but it works 👍

File: Classes/MapEditor/Stats/MapStats.cs (hunk 2 above)
Comment (Brollyy) [reply Classes/MapEditor/Stats/MapStats.cs:110]: This certainly works, but is **very** noticably slower with around 1000 notes. The `foreach` approach has linear time complexity, while the LINQ version is quadratic.

Comment (Brollyy) [reply Classes/MapEditor/Stats/MapStats.cs:110]: Nevermind, I figured out a faster way to do it with `SortedSet` :D

**Review by Nytilde — APPROVED** [review]
looks good to me :)

**Review by PKBeam — DISMISSED** [review]

**Review by PKBeam — APPROVED** [review]

**Pull request merged**